metrics/sinks/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,190 @@
+	if len(uri.Host) > 0 {
+		return nil, fmt.Errorf("Host should not be set for GKE sink")
+	}
+
+	if err := gce_util.EnsureOnGCE(); err != nil {
+		return nil, err
+	}
+
+	// Detect project ID
+	projectId, err := gce.ProjectID()
+	if err != nil {
+		return nil, err
+	}
+
+	// Detect zone
+	zone, err := gce.Zone()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create Google Cloud Monitoring service
+	client := oauth2.NewClient(oauth2.NoContext, google.ComputeTokenSource(""))
+	stackdriverClient, err := sd_api.New(client)
+	if err != nil {
+		return nil, err
+	}
+
+	sink := &stackdriverSink{
+		project:           projectId,
+		zone:              zone,
+		stackdriverClient: stackdriverClient,
+	}
+
+	glog.Infof("Created GKE sink")
+
+	return sink, nil
+}
+
+func (sink *stackdriverSink) sendRequest(req *sd_api.CreateTimeSeriesRequest) {
+	_, err := sink.stackdriverClient.Projects.TimeSeries.Create(fullProjectName(sink.project), req).Do()
+	if err != nil {
+		glog.Errorf("Error while sending request to Stackdriver %v", err)
+	} else {
+		glog.V(4).Infof("Successfully sent %v timeseries to Stackdriver", len(req.TimeSeries), sink.project)
+	}
+}
+
+func (sink *stackdriverSink) translateMetric(timestamp time.Time, labels map[string]string, name string, value core.MetricValue, createTime time.Time) *sd_api.TimeSeries {
+	switch name {
+	case core.MetricUptime.MetricDescriptor.Name:
+		point := sink.uptimePoint(timestamp, createTime, value)
+		resourceLabels := sink.getResourceLabels(labels)
+		return createTimeSeries(resourceLabels, labels, uptimeMD, point)
+	default:
+		return nil
+	}
+}
+
+func (sink *stackdriverSink) getResourceLabels(labels map[string]string) map[string]string {
+	return map[string]string{
+		"project_id":     sink.project,
+		"cluster_name":   "",
+		"zone":           sink.zone,
+		"instance_id":    labels[core.LabelHostID.Key],
+		"namespace_id":   labels[core.LabelPodNamespaceUID.Key],
+		"pod_id":         labels[core.LabelPodId.Key],
+		"container_name": labels[core.LabelContainerName.Key],
+	}
+}
+
+func createTimeSeries(resourceLabels map[string]string, labels map[string]string, metadata *metricMetadata, point *sd_api.Point) *sd_api.TimeSeries {
+	return &sd_api.TimeSeries{
+		Metric: &sd_api.Metric{
+			Labels: map[string]string{},
+			Type:   metadata.Name,
+		},
+		MetricKind: metadata.MetricKind,
+		ValueType:  metadata.ValueType,
+		Resource: &sd_api.MonitoredResource{
+			Labels: resourceLabels,
+			Type:   "stackdriver_container",
+		},
+		Points: []*sd_api.Point{point},
+	}
+}
+
+func (sink *stackdriverSink) uptimePoint(timestamp time.Time, createTime time.Time, value core.MetricValue) *sd_api.Point {
+	return &sd_api.Point{
+		Interval: &sd_api.TimeInterval{
+			EndTime:   timestamp.Format(time.RFC3339),
+			StartTime: createTime.Format(time.RFC3339),
+		},
+		Value: &sd_api.TypedValue{
+			DoubleValue:     float64(value.IntValue) / float64(time.Second/time.Millisecond),
+			ForceSendFields: []string{"DoubleValue"},
```

Comment:
Not adding this field caused custom metrics to be rejected by Stackdriver. I did a test and writing GKE container metrics seems to work without it, removing this (although kubelet-to-gcm still uses it?)